src/coreclr/vm/threadsuspend.cpp

```diff
@@ -3209,6 +3210,44 @@ COR_PRF_SUSPEND_REASON GCSuspendReasonToProfSuspendReason(ThreadSuspend::SUSPEND
 }
 #endif // PROFILING_SUPPORTED
 
+static int64_t QueryPerformanceCounter()
+{
+    LARGE_INTEGER ts;
+    QueryPerformanceCounter(&ts);
+    return ts.QuadPart;
+}
+
+static int64_t QueryPerformanceFrequency()
+{
+    LARGE_INTEGER ts;
+    QueryPerformanceFrequency(&ts);
+    return ts.QuadPart;
+}
+
+// exponential spinwait with an approximate time limit for waiting in microsecond range.
+// when iteration == -1, only usecLimit is used
+void SpinWait(int iteration, int usecLimit)
```

Comment:
is this existing logic or a new mechanism? Changing the spinwait duration has shown to impact startup when DATAS is enabled.

Reply:
This is not new, it is copied from NativeAOT. It is not related to spinwaits used by GC in any way. Here we just need to keep the thread that performs the suspension busy for a few microseconds. System timers do not offer pauses with such granularity, not in a portable way, anyways.

Reply:
What is going on here: 
- We set a flag telling threads to suspend themselves. (post a frame suitable for stackwalk start and park on an event).
- Most threads do some kind of allocations or calls into runtime, so they will notice the flag and suspend.
- Some threads may not be doing that (computing something in a loop), so they need to be hijacked.

Hijack may either catch a thread in interruptible code (then we are done with the thread) or hijack the return address, so that when a thread returns from current call, it will suspend itself.

Hijack typically leads to thread suspension, but not always (a thread may go deeper into call tree), so we might need to redo the hijack a few times. We should eventually "corner" the thread as hijacked return will be moving only lower in the call tree with every try.  
Here is a conundrum though - after hijacking we must let the thread to run for a while, so it has a chance to observe what we did and suspend itself. We can't be too aggressive with this. If we keep interrupting the thread to check what is happening, nothing will happen. So after a hijack cycle we back off for a few microseconds and then check if we need to hijack again.

We will also increase the time we give to the thread with every iteration (to make sure we are not starving it with our interruptions), but up to a limit. 

Timings here are derived from the general pause expectations - 1/60 second (16 msec) pause could be perceptible, 1/15 second is certainly noticeable in interactive apps. But this is just the suspension part, we need to leave some time for the GC to run. While we can't guarantee the upper bound, we strive for suspension to happen in sub-millisecond time. 
Thus the retry iterations are measured in microseconds.